core/pta/rockchip/rk_secure_boot.c

```diff
@@ -0,0 +1,406 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2025, Pengutronix, Michael Tretter <entwicklung@pengutronix.de>
+ */
+
+#include <config.h>
+#include <drivers/rockchip_otp.h>
+#include <kernel/pseudo_ta.h>
+#include <kernel/tee_misc.h>
+#include <tee/uuid.h>
+#include <utee_defines.h>
+#include <stdio.h>
+#include <string.h>
+#include <platform_config.h>
+
+#include <pta_rk_secure_boot.h>
+
+#define PTA_NAME "rk_secure_boot.pta"
+
+/*
+ * Disable simulation to actually fusing the hash into the OTP. Enabling this
+ * option is necessary to actually enable secure boot with PTA, but may
+ * potentially brick your device.
+ */
+static const int simulation = IS_ENABLED2(CFG_RK_SECURE_BOOT_SIMULATION);
+
+static inline bool bit_test(uint32_t value, uint32_t bit)
+{
+	return (value & bit) == bit;
+}
+
+/* Convert hash from 32 bit words in OTP into a byte array */
+static TEE_Result otp_to_bytes(uint32_t *otp, size_t otp_size,
+			       uint8_t *bytes, size_t bytes_size)
+{
+	uint32_t *u32 = otp;
+	uint8_t *u8 = bytes;
+	size_t i;
+	size_t j;
+
+	if (bytes_size * sizeof(*bytes) != otp_size * sizeof(*otp))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < otp_size; i++, u32++)
+		for (j = 0; j < sizeof(*u32); j++, u8++)
+			*u8 = (uint8_t)(*u32 >> (j * 8));
+
+	return TEE_SUCCESS;
+}
+
+/* Convert hash from byte array into 32 bit words for OTP */
+static TEE_Result bytes_to_otp(uint8_t *bytes, size_t bytes_size,
+			       uint32_t *otp, size_t otp_size)
+{
+	uint32_t *u32 = otp;
+	uint8_t *u8 = bytes;
+	size_t i;
+	size_t j;
+
+	if (bytes_size * sizeof(*bytes) != otp_size * sizeof(*otp))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < otp_size; i++, u32++) {
+		*u32 = 0;
+		for (j = 0; j < sizeof(*u32); j++, u8++)
+			*u32 += (uint32_t)*u8 << (j * 8);
+	}
+
+	return TEE_SUCCESS;
+}
+
+#define HASH_STRING_SIZE 88
+static TEE_Result otp_to_string(uint32_t *otp, size_t otp_size,
+				char *str, size_t str_size)
+{
+	if (otp_size != 8 || str_size != HASH_STRING_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	snprintf(str, str_size,
+		 "0x%08x 0x%08x 0x%08x 0x%08x 0x%08x 0x%08x 0x%08x 0x%08x",
+		 otp[0], otp[1], otp[2], otp[3],
+		 otp[4], otp[5], otp[6], otp[7]);
+
+	return TEE_SUCCESS;
+}
```

Comment:
I wanted to be able to decide the log level in the caller. Thus, I implemented this as a format helper, but it's still clumsy to use. Unfortunately, I didn't find any better helpers in OP-TEE. Did I miss something obvious, or is there a better way to solve this?

Reply:
You never check the return value from otp_to_string() and the you're always supposed to use a proper otp hash size as input. How about simplifying it to allow it to be called in an argument to the *MSG() macro like:
``` C
static char *otp_to_string(uint32_t *otp, char *str, size_t str_size)
{
        snprintf(str, str_size, ...);

        return str;
}
```